stan/math/rev/mat/fun/cov_per.hpp

```diff
@@ -0,0 +1,347 @@
+     * @tparam T_p type of period
+     */
+    template <typename T_x, typename T_sigma, typename T_l, typename T_p>
+    class cov_per_vari : public vari {
+    public:
+      const size_t size_;
+      const size_t size_ltri_;
+      const double l_d_;
+      const double sigma_d_;
+      const double p_d_;
+      const double sigma_sq_d_;
+      double* dist_;
+      double* sin_dist_;
+      double* cos_dist_;
+      double* sin_dist_sq_;
+      vari* l_vari_;
+      vari* sigma_vari_;
+      vari* p_vari_;
+      vari** cov_lower_;
+      vari** cov_diag_;
+
+      /**
+       * Constructor for cov_per.
+       *
+       * All memory allocated in
+       * ChainableStack's stack_alloc arena.
+       *
+       * It is critical for the efficiency of this object
+       * that the constructor create new varis that aren't
+       * popped onto the var_stack_, but rather are
+       * popped onto the var_nochain_stack_. This is
+       * controlled to the second argument to
+       * vari's constructor.
+       *
+       * @param x std::vector input that can be used in square distance
+       *    Assumes each element of x is the same size
+       * @param sigma standard deviation
+       * @param l length scale
+       * @param p period
+       */
+      cov_per_vari(const std::vector<T_x>& x,
+                        const T_sigma& sigma,
+                        const T_l& l,
+                        const T_p& p
+						)
+        : vari(0.0),
+          size_(x.size()),
+          size_ltri_(size_ * (size_ - 1) / 2),
+          l_d_(value_of(l)), sigma_d_(value_of(sigma)),
+		  p_d_(value_of(p)),
+          sigma_sq_d_(sigma_d_ * sigma_d_),
+		  dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  sin_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  cos_dist_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+		  sin_dist_sq_(ChainableStack::memalloc_.alloc_array<double>(size_ltri_)),
+          l_vari_(l.vi_), sigma_vari_(sigma.vi_),
+		  p_vari_(p.vi_),
+          cov_lower_(ChainableStack::memalloc_.alloc_array<vari*>(size_ltri_)),
+          cov_diag_(ChainableStack::memalloc_.alloc_array<vari*>(size_)) {
+        double neg_two_inv_l_sq = -2.0 / (l_d_ * l_d_);
+        double pi_div_p = M_PI / p_d_;
+
+        size_t pos = 0;
+        for (size_t j = 0; j < size_ - 1; ++j) {
+          for (size_t i = j + 1; i < size_; ++i) {
+        	double dist = distance(x[i], x[j]);
+        	double sin_dist = sin(pi_div_p * dist);
+            double sin_dist_sq = square(sin_dist);
+            dist_[pos] = dist;
+            sin_dist_[pos] = sin_dist;
+        	cos_dist_[pos] = cos(pi_div_p * dist);
+            sin_dist_sq_[pos] = sin_dist_sq;
+            cov_lower_[pos] = new vari(sigma_sq_d_ * std::exp(sin_dist_sq
+                                       * neg_two_inv_l_sq), false);
+            ++pos;
+          }
+        }
+        for (size_t i = 0; i < size_; ++i)
```

Comment:
Same comment about size_ - 1 vs. size_ for j loop to avoid this

Reply:
done